subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformationResult.java

```diff
@@ -0,0 +1,144 @@
+        }
+
+        public TransformationResult build() {
+            ImmutableList<TransformationOutput> transformationOutputs = builder.build();
+            return onlyProducedOutputs
+                ? new OnlyProducedOutputTransformationResult(convertToProducedOutputLocations(transformationOutputs))
+                : new DefaultTransformationResult(transformationOutputs);
+        }
+
+        private interface TransformationOutput {
+            File resolveForInputArtifact(File inputArtifact);
+        }
+
+        private static ImmutableList<File> convertToProducedOutputLocations(ImmutableList<TransformationOutput> transformationOutputs) {
+            ImmutableList.Builder<File> builder = new ImmutableList.Builder<>();
+            transformationOutputs.forEach(output -> builder.add(((ProducedOutput) output).getOutputLocation()));
+            return builder.build();
+        }
+
+        private static class OnlyProducedOutputTransformationResult implements TransformationResult {
+            private final ImmutableList<File> producedOutputLocations;
+
+            public OnlyProducedOutputTransformationResult(ImmutableList<File> producedOutputLocations) {
+                this.producedOutputLocations = producedOutputLocations;
+            }
+
+            @Override
+            public ImmutableList<File> resolveOutputsForInputArtifact(File inputArtifact) {
+                return producedOutputLocations;
+            }
+        }
+
+        private static class DefaultTransformationResult implements TransformationResult {
+            private final ImmutableList<TransformationOutput> transformationOutputs;
+
+            public DefaultTransformationResult(ImmutableList<TransformationOutput> transformationOutputs) {
+                this.transformationOutputs = transformationOutputs;
+            }
+
+            @Override
+            public ImmutableList<File> resolveOutputsForInputArtifact(File inputArtifact) {
+                ImmutableList.Builder<File> builder = ImmutableList.builderWithExpectedSize(transformationOutputs.size());
+                transformationOutputs.forEach(output -> builder.add(output.resolveForInputArtifact(inputArtifact)));
+                return builder.build();
+            }
+        }
+
+        private static class InInputArtifact implements TransformationOutput {
+            private final String relativePath;
+
+            public InInputArtifact(String relativePath) {
+                this.relativePath = relativePath;
+            }
+
+            @Override
+            public File resolveForInputArtifact(File inputArtifact) {
+                return new File(inputArtifact, relativePath);
+            }
+        }
+
+        private static class InputArtifact implements TransformationOutput {
+            @Override
+            public File resolveForInputArtifact(File inputArtifact) {
+                return inputArtifact;
+            }
+        }
```

Comment:
I think something like `EntireInputArtifact` and `PartOfInputArtifact` or `DescendantOfInputArtifact` or `FragmentOfInputArtifact` would be a better distinction.

Reply:
Renamed.